File: mvp/src/main/java/com/github/wally/mvp/http/cache/Repository.java
package com.github.wally.mvp.http.cache;

import com.github.wally.mvp.http.api.GankApiService;
import com.github.wally.mvp.bean.gank.GankMeiZiListBean;
import com.github.wally.mvp.http.RetrofitManager;

import java.io.File;

import io.reactivex.Observable;
import io.rx_cache2.DynamicKey;
import io.rx_cache2.internal.RxCache;
import io.victoralbertos.jolyglot.GsonSpeaker;

/**
 * Package: com.github.wally.mvp.http.cache
 * FileName: Repository
 * Date: on 2018/6/16  下午7:21
 * Auther: zihe
 * Descirbe:
 * Email: hezihao@linghit.com
 */
public class Repository {
    private final GankCacheProviders cacheProviders;
    private final GankApiService mGankApiService;

    public static Repository init(File cacheDir) {
        return new Repository(cacheDir);
    }

    public Repository(File cacheDir) {
        cacheProviders = new RxCache.Builder()
                .persistence(cacheDir, new GsonSpeaker())
                .using(GankCacheProviders.class);

        mGankApiService = RetrofitManager.getInstance().getGankApiService();
    }

    public Observable<GankMeiZiListBean> getMeizi(int size, int num) {
        return cacheProviders.getMeizi(mGankApiService.getMeizi(size, num), new DynamicKey(num));
    }
}

File: gank/src/main/java/com/github/wally/gank/http/cache/Repository.kt
package com.github.wally.gank.http.cache

import com.github.wally.gank.bean.gank.GankMeiZiListBean
import com.github.wally.gank.bean.gank.GankSearchBean
import com.github.wally.gank.http.RetrofitManager
import com.github.wally.gank.http.api.GankApiService
import io.reactivex.Observable
import io.rx_cache2.DynamicKey
import io.rx_cache2.DynamicKeyGroup
import io.rx_cache2.internal.RxCache
import io.victoralbertos.jolyglot.GsonSpeaker
import java.io.File

/**
 * Package: com.github.wally.gank.http.cache
 * FileName: Repository
 * Date: on 2018/6/16  下午7:21
 * Auther: zihe
 * Descirbe:
 * Email: hezihao@linghit.com
 */
class Repository(cacheDir: File) {
    private val cacheProviders: GankCacheProviders
    private val mGankApiService: GankApiService?

    init {
        cacheProviders = RxCache.Builder()
                .persistence(cacheDir, GsonSpeaker())
                .using(GankCacheProviders::class.java)
        mGankApiService = RetrofitManager.instance.gankApiService
    }

    /**
     * 获取妹子图列表
     *
     * @param size 请求多少条
     * @param num  请求的页码
     * @return 数据源
     */
    fun getMeizi(size: Int, num: Int): Observable<GankMeiZiListBean> {
        return cacheProviders.getMeizi(mGankApiService!!.getMeizi(size, num), DynamicKey(num))
    }

    /**
     * 获取搜索分类列表
     *
     * @param category 分类的名字
     * @param count    请求多少条
     * @param page     请求的页码
     * @return 数据源
     */
    fun getSearchCategoryList(category: String, count: Int, page: Int): Observable<GankSearchBean> {
        return cacheProviders.getSearchCategoryList(mGankApiService!!.getSearchCategoryList(category, count, page), DynamicKeyGroup(page, category))
    }

    companion object {
        fun init(cacheDir: File): Repository {
            return Repository(cacheDir)
        }
    }
}